clients/drcachesim/tools/view.cpp

```diff
@@ -591,47 +601,12 @@ view_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
         return true;
     }
 
-    // XXX: We could potentially use decode_cache_t here (i#7113) and avoid the
-    // duplicated caching logic.
-    app_pc decode_pc;
-    const app_pc orig_pc = (app_pc)memref.instr.addr;
-    if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, filetype_)) {
-        // The trace has instruction encodings inside it.
-        decode_pc = const_cast<app_pc>(memref.instr.encoding);
-        if (memref.instr.encoding_is_new) {
-            // The code may have changed: invalidate the cache.
-            disasm_cache_.erase(orig_pc);
-        }
-    } else {
-        // Legacy trace support where we need the binaries.
-        decode_pc = module_mapper_->find_mapped_trace_address(orig_pc);
-        if (!module_mapper_->get_last_error().empty()) {
-            error_string_ = "Failed to find mapped address for " +
-                to_hex_string(memref.instr.addr) + ": " +
-                module_mapper_->get_last_error();
-            return false;
-        }
-    }
+    disasm_info_t *disasm_info;
+    error_string_ = decode_cache_->add_decode_info(memref.instr, disasm_info);
+    if (!error_string_.empty())
+        return false;
+    std::string disasm = disasm_info->disasm_;
```

Comment:
Can we avoid this string copy by directly using disasm_info->disasm_ below?  Is there any reason to make a copy?

Reply:
I don't understand. Various additional decorations are added to `disasm` that are based on the exact type of memref in this instance. We don't want those decorations to be part of the cached information for this pc.

Are you suggesting that the cache should use `(pc, memref_instr_type)` as key so we don't need to recompute those decorations?

Reply:
I can see a way to do it but it would either involve duplicating some of the below conditions, or generally reducing readability of this snippet. Can be done, but is it worth avoiding this copy for a non-performance-sensitive tool that is anyway limited by file/console writing overhead?

Reply:
Extra overhead here can outweigh the i/o overhead: without caching, the decode + disassembly time dominates IIRC.

I would think a string pointer would allow avoiding the copy without any duplication or loss of readability.  But probably you are right that it won't help perf; esp if you have measured it and it's the i/o that dominates.

Reply:
perf says:
- 11% time is spend in std::ostream related functions,
- 6% spent in decode_cache_t,
- 0.144% in basic_string::operator=

Reply:
> 0.144% in basic_string::operator=

This is all instances of string assignments, not just the one we're discussing about.